### DISCUSSION TITLE
Any experiences with React Flow and WebAssembly?

### DISCUSSION THREAD

**Discussion by aschrijver** [xyflow/xyflow#3148] [Q&A]
I am evaluating WebAssembly for use both in the browser and on the server. This allows among others doing full-stack development in a language traditionally not for the web. In my case that's Rust, and I am looking into the [Leptos](https://leptos.dev/) reactive web framework in particular (it's similar to SolidJS and significantly more performant ([video](https://yewtu.be/watch?v=4KtotxNAwME)) than React webapps).

But I want to have diagramming capabilities in the Leptos app, and React Flow is by far the best project I can find. However, I can't find a lot of good information on how to use NPM modules (like React Flow) in WebAssembly-enabled web frameworks. Does anyone have experience in doing so? And are there plans / interest to create a WebAssembly module or component that wraps React Flow, so that it is ready for use in these environments?

Because the WebAssembly ecosystem is broad, there are several related discussions I created:

- https://github.com/leptos-rs/leptos/discussions/1206
- https://github.com/rustwasm/wasm-bindgen/discussions/3488
- https://github.com/WasmEdge/WasmEdge/discussions/2607

**Comment by coopbri** [comment]
Hey @aschrijver, thanks for making all of these related issues. I am trying to do much the same and I've just begun exploring the discussions you linked. I noticed you mentioned Extism, which is a project I have been using as well to try to get this (and similar WASM/WASI/web projects) working. Maybe we can collaborate, not sure if you came up with a clean solution to make this work yet. Feel free to reach out to me on Twitter, Discord, or LinkedIn (all linked on my GitHub profile) if you'd like to chat!

**Reply by aschrijver** [reply]
Hi @coopbri :wave: Wasm is fascinating and getting uptake. I am currently looking into [wasmCloud](https://wasmcloud.com) in combination with [Concordance](https://github.com/cosmonic/concordance) (CQRS/ES). Esp. the latter is still pre-MVP, but in latest [community meeting](https://wasmcloud.com/community/2023/08/16/community-meeting) @autodidaddict demonstrated an integration with [Event Catalog](https://www.eventcatalog.dev/) to get a kind of document-first approach to development (important when doing domain-driven design).

Now Event Catalog is nice, and it has MermaidJS integration. Their approach led me a bit deeper and finding another diagramming candidate, namely [D2 Language](https://d2lang.com/), which is a more modern diagrams-as-code approach wrt Mermaid / PlantUML. D2 has an [open PR](https://github.com/terrastruct/d2/pull/436) to compile to Wasm, and has an API that'd allow similar kind of dynamic interaction as React Flow.

So, I'm not sure yet what would be the better approach for me, React Flow vs. D2 + Gui front-end.

Also on Rust GUI subject, I'm a bit hesitant. Web-based frameworks using the DOM should be mostly fine, i.e. in terms of a11y and i18n. There's this [HN submission](https://news.ycombinator.com/item?id=35722681) contrasting a bunch of the approaches. In terms of GUI I might better take a Tauri + Vite + React safe choice, and then React Flow integration is straightforward.

As for polyglot plugin architecture Extism has a headstart wrt wasmCloud. Different positioning. With wasmCloud event-driven, actor-based approach I should be fine, though.

**Reply by coopbri** [reply]
Thank you so much @aschrijver! These are great leads. I hadn't heard of wasmCloud nor its Concordance event framework yet. Hadn't heard of D2 yet either, but big fan of Mermaid. Also have had a great experience with Tauri (+ even [Tauri mobile](https://github.com/tauri-apps/tauri-mobile), which is very alpha) personally.

Very turbulent space in WASM right now, really excited to see it play out.

Cheers

**Reply by moklick** [reply]
Thanks for sharing @aschrijver . D2 looks super interesting 🧐
